modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/DocumentLibrary/DocumentLibrary.es.js

```diff
@@ -16,310 +16,280 @@ import '../FieldBase/FieldBase.es';
-			p_p_mode: 'view',
-			p_p_state: 'pop_up',
-			refererGroupId: themeDisplay.getScopeGroupId(),
-		};
-
-		const documentLibrarySelectorURL = createPortletURL(
-			themeDisplay.getLayoutRelativeControlPanelURL(),
-			documentLibrarySelectorParameters
-		);
+	return [fileEntryTitle, fileEntryURL];
+}
 
-		return documentLibrarySelectorURL.toString();
-	}
+const DocumentLibrary = ({
+	fileEntryTitle = '',
+	fileEntryURL = '',
+	id,
+	name,
+	onClearButtonClicked,
+	onSelectButtonClicked,
+	placeholder,
+	readOnly,
+	value,
+}) => {
+	const [transformedFileEntryTitle, transformedFileEntryURL] = useMemo(
+		() =>
+			transformFileEntryProperties({
+				fileEntryTitle,
+				fileEntryURL,
+				value,
+			}),
+		[fileEntryTitle, fileEntryURL, value]
+	);
+
+	return (
+		<div className="liferay-ddm-form-field-document-library">
+			{transformedFileEntryURL && readOnly ? (
+				<CardItem fileEntryTitle={transformedFileEntryTitle} />
+			) : (
+				<ClayInput.Group>
+					<ClayInput.GroupItem prepend>
+						<ClayInput
+							aria-label={Liferay.Language.get('file')}
+							className="field"
+							disabled
+							id={`${name}inputFile`}
+							value={transformedFileEntryTitle}
+						/>
+					</ClayInput.GroupItem>
+
+					<ClayInput.GroupItem append shrink>
+						<ClayButton
+							className="select-button"
+							disabled={readOnly}
+							displayType="secondary"
+							onClick={onSelectButtonClicked}
+						>
+							<span className="lfr-btn-label">
+								{Liferay.Language.get('select')}
+							</span>
+						</ClayButton>
+					</ClayInput.GroupItem>
+
+					{transformedFileEntryTitle && (
+						<ClayInput.GroupItem append shrink>
+							<ClayButtonWithIcon
+								className="clear-button"
+								displayType="secondary"
+								onClick={onClearButtonClicked}
+								symbol="times"
+							/>
+						</ClayInput.GroupItem>
+					)}
+				</ClayInput.Group>
+			)}
+
+			<ClayInput
+				id={id}
+				name={name}
+				placeholder={placeholder}
+				type="hidden"
+				value={transformedFileEntryTitle}
+			/>
+		</div>
+	);
+};
 
-	getUploadURL() {
-		const uploadParameters = {
-			cmd: 'add_temp',
-			'javax.portlet.action': '/document_library/upload_file_entry',
-			p_auth: Liferay.authToken,
-			p_p_id: Liferay.PortletKeys.DOCUMENT_LIBRARY,
-		};
-
-		const uploadURL = createActionURL(
-			themeDisplay.getLayoutRelativeURL(),
-			uploadParameters
+const DocumentLibraryProxy = connectStore(
+	({
+		emit,
+		fileEntryTitle,
+		fileEntryURL,
+		itemSelectorAuthToken,
+		store,
+		value = {},
+		...otherProps
+	}) => {
+		const [currentValue, setCurrentValue] = useState(value);
+
+		const _handleVisibleChange = useCallback(
+			event => {
+				if (event.selectedItem) {
+					emit('fieldFocused', event, event.selectedItem);
+				}
+				else {
+					emit('fieldBlurred', event);
+				}
+			},
+			[emit]
 		);
 
-		return uploadURL.toString();
-	}
+		const _handleSelectButtonClicked = useCallback(
```

Comment:
We don't need `useCallback` in these functions, we can remove it in these functions, we have no problem losing reference every time it causes a new rendering, we have no other elements that need equal reference during renderings.

Reply:
Ok!